A few forward AD formulas

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #77421

It includes all-time favourites like:
- `put`
- `nn.functional.embedding`
- `prelu`
- `nn.functional.bilinear`
- `nn.functional.rrelu`
- `nn.functional.logsigmoid`

lezcano committed:
A few forward AD formulas

It includes all-time favourites like:
- `put`
- `nn.functional.embedding`
- `prelu`
- `nn.functional.bilinear`
- `nn.functional.rrelu`
- `nn.functional.logsigmoid`

[ghstack-poisoned]

Lezcano commented on aten/src/ATen/native/Activation.cpp:
```diff
@@ -588,14 +588,13 @@ Tensor rrelu_with_noise_backward(
+  if (training) {
+    return noise * grad_output;
   } else {
-    auto negative = (lower_tensor + upper_tensor) / 2;
-    Scalar negative_slope = negative.item();
-    return at::leaky_relu_backward(grad_output, self_or_result, negative_slope, is_result);
+    auto l = lower.toDouble();
+    auto u = upper.toDouble();
+    auto mid = (l + u) / 2.;
+    return at::leaky_relu_backward(grad_output, self_or_result, mid, is_result);
```
This is just a simplification of the previous formula (which was incorrect when `lower > upper` I believe) that avoids using and inspecting the elements of tensors.

soulitzer commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -13240,6 +13240,13 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
+            # In-place operations do not play well with forward AD
+            # FIXME it should be fixable, as auto_elementwise works with in-place just fine, and
+            # this one is somewhat similar
+            # INTERNAL ASSERT FAILED Cannot set a va lue of a forward grad if it already exists. 
```
Could you file an issue for this?

Lezcano replied:
https://github.com/pytorch/pytorch/issues/77447

soulitzer reviewed (approved):
Thanks!

lezcano committed:
Update on "A few forward AD formulas"


It includes all-time favourites like:
- `put`
- `nn.functional.embedding`
- `prelu`
- `nn.functional.bilinear`
- `nn.functional.rrelu`
- `nn.functional.logsigmoid`

[ghstack-poisoned]

Lezcano commented:
@pytorchmergebot merge on green

lezcano committed:
Update on "A few forward AD formulas"


It includes all-time favourites like:
- `put`
- `nn.functional.embedding`
- `prelu`
- `nn.functional.bilinear`
- `nn.functional.rrelu`
- `nn.functional.logsigmoid`

[ghstack-poisoned]
